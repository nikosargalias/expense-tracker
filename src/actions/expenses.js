import generateNewId from "../helpers/generateId";
import store from "../store/configureStore";

// action generator
const addExpense = (expense) => {
  const prevState = store.getState();
  return {
    type: "ADD_EXPENSE",
    expense: {
      description: "",
      note: "",
      amount: 0,
      createdAt: new Date().getTime(),
      ...expense,
      id: generateNewId(prevState.expenses),
    },
  };
};

const removeExpense = (id) => {
  return {
    type: "REMOVE_EXPENSE",
    id,
  };
};

const editExpense = ({ id, updates } = {}) => {
  console.log(id, updates);
  return {
    type: "EDIT_EXPENSE",
    id,
    updates,
  };
};

export { addExpense, removeExpense, editExpense };
