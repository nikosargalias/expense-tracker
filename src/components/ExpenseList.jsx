import React from "react";
import { connect } from "react-redux";
import ExpenseListItem from "./ExpenseListItem";
import selectExpenses from "../selectors/expenses";
import ExpenseListFilters from "./ExpenseListFilters";

export const ExpenseList = (props) => {
  return (
    <div>
      <h1>Expense List</h1>
      <ExpenseListFilters />
      {(props.expenses && (
        <ul>
          {props.expenses.map((expense, i) => (
            <ExpenseListItem {...expense} key={expense.id} />
          ))}
        </ul>
      )) || <p>No expenses </p>}
    </div>
  );
};

const mapStateToProps = (state) => {
  return { expenses: selectExpenses(state.expenses, state.filters) };
};

export default connect(mapStateToProps)(ExpenseList);
