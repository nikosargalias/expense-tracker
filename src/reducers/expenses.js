const expensesReducerDefaultState = [];

const expenseReducer = (
  prevState = expensesReducerDefaultState,
  { type, expense, id, updates }
) => {
  switch (type) {
    case "ADD_EXPENSE":
      return [...prevState, expense];
    case "REMOVE_EXPENSE":
      return prevState.filter((expense) => expense.id !== id);
    case "EDIT_EXPENSE":
      return prevState.map((expense) => {
        console.log(id);
        if (expense.id === id) {
          return {
            ...expense,
            ...updates,
          };
        }
        return expense;
      });
    default:
      return prevState;
  }
};

export { expenseReducer as default };
