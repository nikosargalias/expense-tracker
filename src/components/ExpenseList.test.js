import React from "react";
import { shallow } from "enzyme";
import { ExpenseList } from "./ExpenseList";
import expenses from "../fixtures/expenses";

test("should render Expense List correctly", () => {
  const wrapper = shallow(<ExpenseList expenses={expenses} />);
  expect(wrapper).toMatchSnapshot();
});

test("should render Expense list with no expenses message", () => {
  const wrapper = shallow(<ExpenseList />);
  expect(wrapper).toMatchSnapshot();
});
