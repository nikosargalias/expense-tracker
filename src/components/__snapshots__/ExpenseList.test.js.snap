// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Expense List correctly 1`] = `
<div>
  <h1>
    Expense List
  </h1>
  <Connect(ExpenseListFilters) />
  <ul>
    <Connect(ExpenseListItem)
      amount={2000}
      createdAt={0}
      description="gum"
      id={1}
      key="1"
      note="Chew"
    />
    <Connect(ExpenseListItem)
      amount={100}
      createdAt={-259200000}
      description="drink"
      id={2}
      key="2"
      note="nothing"
    />
    <Connect(ExpenseListItem)
      amount={4500}
      createdAt={172800000}
      description="Rent"
      id={3}
      key="3"
      note="Expensive"
    />
  </ul>
</div>
`;

exports[`should render Expense list with no expenses message 1`] = `
<div>
  <h1>
    Expense List
  </h1>
  <Connect(ExpenseListFilters) />
  <p>
    No expenses 
  </p>
</div>
`;
